File: src-py/speedTest.py
from led_manager import set_color
import time
from simpleLog import log
from random import randint

numLed = 100

# log("All white!")
# for i in range(numLed):
#     set_color(i, 0, 0, 255)
# time.sleep(1)


# log("All random!")
# for i in range(numLed):
#     set_color(i, randint(0, 255), randint(0, 255), randint(0 ,255))
# time.sleep(1)

# log("Testing 15 sequential loops.") # 750 writes
# start = time.time()
# for i in range (15):
#     for i in range(numLed):
#         set_color(i, randint(0, 255), randint(0, 255), randint(0 ,255))
# end = time.time()
# log(str(end-start)+" seconds. ")
# log(str((end-start)/750)+" seconds per LED.")
# log(str(750/(end-start))+" LEDs per second.\n\n")

log("Testing 750 random writes.") 
start = time.time()
for i in range (750):
    set_color(randint(0, numLed), randint(0, 255), randint(0, 255), randint(0, 255))
end = time.time()
log(str(end-start)+" seconds. ")
log(str((end-start)/750)+" seconds per LED.")
log(str(750/(end-start))+" LEDs per second.")

File: src/speedtest.rs
use std::{
    sync::{Arc, Mutex},
    time::Instant,
};

use log::{debug, info};
use rand::Rng;

use crate::{led_manager, ManagerData};

pub fn speedtest(manager: &Arc<Mutex<ManagerData>>, num_led: u32, writes: u32) {
    let mut rng = rand::thread_rng();
    info!("Clearing string");

    for n in 0..num_led {
        led_manager::set_color(manager, n as u16, 0, 0, 0);
    }

    info!("Testing {writes} random writes");
    let start = Instant::now();

    for _n in 0..=writes {
        led_manager::set_color(
            manager,
            rng.gen_range(0..(num_led as u16)),
            rng.gen_range(0..255),
            rng.gen_range(0..255),
            rng.gen_range(0..255),
        );
    }

    let queue_lengths = manager
        .lock()
        .unwrap()
        .state
        .led_state
        .queue_lengths
        .clone();

    let end = start.elapsed();

    let mut queue_total_lengths: u32 = 0;
    if !queue_lengths.is_empty() {
        for n in queue_lengths.iter().take((queue_lengths.len() - 1) + 1) {
            queue_total_lengths += queue_lengths[*n as usize] as u32;
        }
    }

    info!("{end:.2?} seconds.");
    info!("{:.5?} seconds per LED", end / writes);
    info!(
        "{:.3} LEDs per second",
        (writes as f64 / (end.as_millis() as f64)) * 1000.0
    );

    if !queue_lengths.is_empty() {
        info!(
            "Average queue length: {}",
            queue_total_lengths / queue_lengths.len() as u32
        );
    } else {
        debug!(
            "queue_lengths.len() is 0, check debug logs for average queue lengths from threads."
        );
    }
}
